core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -521,7 +524,7 @@ private[spark] class BlockManager(
     // it needs the merge directories metadata which is provided by the local executor during
     // the registration with the ESS. Therefore, this registration should be prior to
     // the BlockManager registration. See SPARK-39647.
-    if (externalShuffleServiceEnabled) {
+    if (externalShuffleServiceEnabled && isSortShuffle) {
```

Comment:
I think the combination of conditions is unreasonable. 

For example, user has customized a `ShuffleManager`, which is compatible with SortShuffleManager, such as `CustomSortShuffleManager`. Before this pr, it may works well, no matter whether `externalShuffleServiceEnabled` is true or false. After this pr, ,it may not work when `externalShuffleServiceEnabled` is enabled because it is not of `SortShuffleManager` type.

So even if it is really necessary to make some changes, I think the new condition should not be `shuffleManager.isInstanceOf[SortShuffleManager]`, which may lead to some unexpected behavior changes.

Reply:
> 

This is indeed a problem, but there doesn't seem to be a better way to tell if it's a Remote shuffle service.

`def isRssEnable(): Boolean = get("spark.shuffle.manager", "sort").contains("RssShuffleManager")  `  
Judging in this way, forcing the class name to be qualified does not seem to be the best implementation either.

Is it possible to consider adding an interface to ShuffleManager to determine whether it is a Remote shuffle service?